splink/blocking.py

```diff
@@ -88,18 +90,30 @@ def exclude_pairs_generated_by_all_preceding_rules_sql(self):
         if not self.preceding_rules:
             return ""
         or_clauses = [
-            br.exclude_pairs_generated_by_this_rule_sql() for br in self.preceding_rules
+            br.exclude_pairs_generated_by_this_rule_sql for br in self.preceding_rules
         ]
         previous_rules = " OR ".join(or_clauses)
         return f"AND NOT ({previous_rules})"
 
-    @property
-    def salted_blocking_rules(self):
-        if self.salting_partitions == 1:
-            yield self.blocking_rule_sql
-        else:
-            for n in range(self.salting_partitions):
-                yield f"{self.blocking_rule_sql} and ceiling(l.__splink_salt * {self.salting_partitions}) = {n+1}"  # noqa: E501
+    def create_blocked_pairs_sql(
+        self, linker: Linker, where_condition, probability, salt_condition=""
```

Comment:
Do we want `salt_condition` to be an argument?

It could also be added as a property to this class in the constructor.

Then, just call `self.salt_condition` in the sql string.

It also means you can delete some code below in the salting class.

Reply:
I agree this doesn't feel quite right, but I think it's a bit harder than it looks.

In the base `BlockingRule` class we have this in `create_blocked_pairs_sql`:

```python
        sql = f"""
            select
            {sql_select_expr}
            , '{self.match_key}' as match_key
            {probability}
            from {linker._input_tablename_l} as l
            inner join {linker._input_tablename_r} as r
            on
            ({self.blocking_rule_sql} {salt_condition})
            {self.exclude_pairs_generated_by_all_preceding_rules_sql}
            {where_condition}
            """
```
Note `salt_condition` is set to `""` i.e. it's not actually needed in the base class.

BUT, we want to be able to reuse this SQL in `SaltedBlockingRule`

In `SaltedBlockingRule` this SQL statement is re-used multiple times and ` UNION ALL `ed, with a varying `salt_condition`:  `AND ceiling(... salt)` where `salt` changes between the ` UNION ALL` statements

So I think:
- In an ideal world, `salt_condition` would not appear in the SQL string in the `BlockingRule` class because that class shouldn't be concerned about salting at all
- I don't see a way of re-using the SQL in the blockingrule class without it 'knowing about' salting, without e.g. manipulation in `sqlglot`
-  `self.salt_condition` can't be a property, because it varies depending on the salt (it could return string template though)
- It could be a function that takes a `salt` argument.  In that case, in the `BlockingRule` class, it would return `""` irrespective of the argument

Reply:
Another solution would be simply to duplicate the sql code, and not use the
```
 sql = super().create_blocked_pairs_sql(
                linker, where_condition, probability, salting_condition
            )
```

Since the SQL string is small, i think the code duplication could probably be justified, so i'd be happy to do that if you think it's better?